nidx/nidx_relation/src/graph_collector.rs

```diff
@@ -0,0 +1,229 @@
+    selector: NodeSelector,
+    schema: crate::schema::Schema,
+}
+
+pub struct TopUniqueRelationCollector {
+    limit: usize,
+    schema: crate::schema::Schema,
+}
+
+pub struct TopUniqueNodeSegmentCollector {
+    limit: usize,
+    selector: NodeSelector,
+    unique: HashSet<NodeId>,
+    schema: crate::schema::Schema,
+    store_reader: StoreReader,
+}
+
+pub struct TopUniqueRelationSegmentCollector {
+    limit: usize,
+    unique: HashMap<RelationId, TantivyDocument>,
+    schema: crate::schema::Schema,
+    store_reader: StoreReader,
+}
+
+impl TopUniqueNodeCollector {
+    pub fn new(selector: NodeSelector, limit: usize) -> Self {
+        Self {
+            limit,
+            selector,
+            schema: crate::schema::Schema::new(),
+        }
+    }
+}
+
+impl TopUniqueRelationCollector {
+    pub fn new(limit: usize) -> Self {
+        Self {
+            limit,
+            schema: crate::schema::Schema::new(),
+        }
+    }
+}
+
+impl Collector for TopUniqueNodeCollector {
+    type Fruit = HashSet<NodeId>;
+    type Child = TopUniqueNodeSegmentCollector;
+
+    fn requires_scoring(&self) -> bool {
+        true
+    }
+
+    fn for_segment(&self, _segment_local_id: SegmentOrdinal, segment: &SegmentReader) -> tantivy::Result<Self::Child> {
+        Ok(TopUniqueNodeSegmentCollector {
+            limit: self.limit,
+            selector: self.selector,
+            unique: HashSet::new(),
+            store_reader: segment.get_store_reader(STORE_READER_CACHED_BLOCKS)?,
+            schema: self.schema.clone(),
+        })
+    }
+
+    fn merge_fruits(
+        &self,
+        segment_fruits: Vec<<Self::Child as SegmentCollector>::Fruit>,
+    ) -> tantivy::Result<Self::Fruit> {
+        let mut unique = HashSet::new();
+        let mut fruits = segment_fruits.into_iter().flatten();
+        let mut fruit = fruits.next();
+
+        while fruit.is_some() && unique.len() < self.limit {
+            unique.insert(fruit.unwrap());
+            fruit = fruits.next();
+        }
+        Ok(unique)
+    }
+}
+
+impl Collector for TopUniqueRelationCollector {
+    type Fruit = Vec<TantivyDocument>;
+    type Child = TopUniqueRelationSegmentCollector;
+
+    fn requires_scoring(&self) -> bool {
+        true
+    }
+
+    fn for_segment(&self, _segment_local_id: SegmentOrdinal, segment: &SegmentReader) -> tantivy::Result<Self::Child> {
+        Ok(TopUniqueRelationSegmentCollector {
+            limit: self.limit,
+            unique: HashMap::new(),
+            store_reader: segment.get_store_reader(STORE_READER_CACHED_BLOCKS)?,
+            schema: self.schema.clone(),
+        })
+    }
+
+    fn merge_fruits(
+        &self,
+        segment_fruits: Vec<<Self::Child as SegmentCollector>::Fruit>,
+    ) -> tantivy::Result<Self::Fruit> {
+        // 1. Merge de hash maps
+        // 2. Build a vec of documents for unique relations
+
+        let mut unique = HashSet::new();
+        let mut docs = Vec::new();
```

Comment:
Why is this needed? You could just return the `unique` HashSet.

Also typo in the comments.

Reply:
Ups, the comments were for myself. Deleting them.
And we need the docs because we want to serialize the relation metadata in the response. The other option would be doing it in the collector and returning the whole `Relation` instead of parts.